File: src/burp_mcp_lite/tools/list_history.py
"""list_history — browse + filter proxy history with field projection."""

from __future__ import annotations

from typing import Optional, Sequence, Union

from ..filters import apply_filter, build_filter
from ..format.render import (
    ALL_FIELDS,
    DEFAULT_FIELDS,
    render_history_ndjson,
    render_history_table,
)
from ..snapshot import SnapshotStore


async def list_history(
    store: SnapshotStore,
    *,
    limit: int = 20,
    offset: int = 0,
    fields: Optional[Sequence[str]] = None,
    host: Optional[str] = None,
    path: Optional[str] = None,
    method: Union[None, str, Sequence[str]] = None,
    status: Optional[str] = None,
    mime: Optional[str] = None,
    match: Optional[str] = None,
    match_in: str = "response.body",
    order: str = "latest",
    format: str = "text",
    refresh: bool = False,
) -> str:
    snap = await store.get(refresh=refresh)
    fil = build_filter(
        host=host,
        path=path,
        method=method,
        status=status,
        mime=mime,
        match=match,
        match_in=match_in,
    )
    filtered = apply_filter(snap.entries, fil)
    if order == "latest":
        # Newest entries first. Ids stay stable (latest entry has the highest
        # id); only the row order in the rendered table is reversed.
        filtered = list(reversed(filtered))
    total = len(filtered)
    page = filtered[offset : offset + max(0, limit)]

    if fields is None:
        active = list(DEFAULT_FIELDS)
    else:
        active = [f for f in fields if f in ALL_FIELDS]
        if not active:
            active = list(DEFAULT_FIELDS)

    if format == "json":
        return render_history_ndjson(page, active)
    return render_history_table(page, fields=active, total=total, offset=offset)

File: src/tools/list-history.ts
// list_history — browse + filter proxy history with field projection.

import { applyFilter, buildFilter } from "../filters.js";
import {
  ALL_FIELDS,
  DEFAULT_FIELDS,
  type Field,
  renderHistoryNdjson,
  renderHistoryTable,
} from "../format/render.js";
import type { SnapshotStore } from "../snapshot.js";

export interface ListHistoryArgs {
  limit?: number;
  offset?: number;
  fields?: readonly string[];
  host?: string;
  path?: string;
  method?: string | readonly string[];
  status?: string;
  mime?: string;
  match?: string;
  matchIn?: string;
  order?: "latest" | "oldest";
  format?: "text" | "json";
  refresh?: boolean;
}

function isField(f: string): f is Field {
  return ALL_FIELDS.has(f as Field);
}

export async function listHistory(
  store: SnapshotStore,
  args: ListHistoryArgs = {},
): Promise<string> {
  const limit = args.limit ?? 20;
  const offset = args.offset ?? 0;
  const order = args.order ?? "latest";
  const format = args.format ?? "text";

  const snap = await store.get(args.refresh ?? false);
  const f = buildFilter({
    host: args.host,
    path: args.path,
    method: args.method,
    status: args.status,
    mime: args.mime,
    match: args.match,
    matchIn: args.matchIn,
  });

  let filtered = applyFilter(snap.entries, f);
  if (order === "latest") filtered = [...filtered].reverse();

  const total = filtered.length;
  const page = filtered.slice(offset, offset + Math.max(0, limit));

  let active: Field[];
  if (args.fields === undefined) {
    active = [...DEFAULT_FIELDS];
  } else {
    active = args.fields.filter(isField);
    if (active.length === 0) active = [...DEFAULT_FIELDS];
  }

  if (format === "json") return renderHistoryNdjson(page, active);
  return renderHistoryTable(page, active, total, offset);
}
